### PR TITLE
Add repo:contains.commit.after() predicate

### PR DESCRIPTION
**Pull request by camdencheek**
Currently, the predicate just acts as an alias that expands to
repohascommitafter since we don't have a way to run arbitrary predicate
code to expand predicates (they must produce a valid query).

### CODE REVIEW

File: internal/search/query/parser.go [hunk 1]
```diff
@@ -434,12 +434,17 @@ func ScanPredicate(field string, buf []byte) (string, int, bool) {
 func ScanPredicateName(fieldRegistry map[string]func() Predicate, buf []byte) (string, int, bool) {
 	var predicateName string
 	var advance int
-	for i, c := range buf {
-		if !unicode.IsLetter(rune(c)) {
-			predicateName = string(buf[:i])
-			advance = i
+	for {
+		r, i := utf8.DecodeRune(buf[advance:])
```
Comment (camdencheek) [inline internal/search/query/parser.go:438]: Small unrelated change to decode runes properly rather than just cast bytes as ASCII to runes.

File: internal/search/query/predicate.go [hunk 2]
```diff
@@ -144,6 +143,33 @@ func (f *RepoContainsPredicate) Plan(parent Basic) (Plan, error) {
...
+
+func (f RepoContainsCommitAfterPredicate) Field() string { return FieldRepo }
+func (f RepoContainsCommitAfterPredicate) Name() string {
+	return "contains.commit.after"
+}
+func (f *RepoContainsCommitAfterPredicate) Plan(parent Basic) (Plan, error) {
+	nodes := make([]Node, 0, 3)
+	nodes = append(nodes, Parameter{
+		Field: FieldCount,
+		Value: "99999",
```
Comment (rvantonder) [inline internal/search/query/predicate.go:163]: Thoughts on whether should inherit count from parent query, or otherwise use default (500 for streaming search)?

**Review by rvantonder — APPROVED** [review]

File: internal/search/query/predicate.go [hunk 3]
```diff
@@ -144,6 +143,33 @@ func (f *RepoContainsPredicate) Plan(parent Basic) (Plan, error) {
 	return ToPlan(Dnf(nodes))
 }
 
+type RepoContainsCommitAfterPredicate struct {
+	TimeRef string
+}
+
+func (f *RepoContainsCommitAfterPredicate) ParseParams(params string) error {
+	f.TimeRef = params
```
[the remark this answers is not in the document]
Comment (camdencheek) [reply internal/search/query/predicate.go:151]: This is the part where I looked into validating it, and it's pretty much impossible without calling out to git unless we restrict what would be allowed here

[the remark this answers is not in the document]
Comment (camdencheek) [reply internal/search/query/predicate.go:151]: See: https://github.com/git/git/blob/master/Documentation/date-formats.txt

File: internal/search/query/predicate.go (hunk 2 above)
Comment (camdencheek) [reply internal/search/query/predicate.go:163]: The problem with using the parent query is say we do something like `repo:contains.commit.after(yesterday) file:go.mod count:1`, the repo filter might hit the limit, return one result, which is then used for the `file:go.mod` query and doesn't match, yielding 0 results even though there are actual matches. 

This is where we get into the "retry with increased count" territory that we do with `AND` queries right now, but we haven't implemented that yet for predicates. At this point, I _think_ we'd rather potentially do extra work and return correct results than return incorrect results. Feel free to disagree though.

File: internal/search/query/predicate.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rvantonder) [reply internal/search/query/predicate.go:151]: Two thoughts:

(1) Still unclear what happens when this is invalid, do we correctly propagate the error up from the git command? Just don't want weird suppression/crashing going on because of the assumption that this value is sane.  Maybe an integration test would assuage my concerns here.

(2) I'm wondering if we might constrain the date format here, if we can't validate by calling out to git here. It seems overkill to just accept any value and have something downstream figure it out. I'm wondering if Go's date format is an acceptable subset, or we can take the Go format as a wrapper and convert it to the ISO spec that will always work with git? This seems a lot more sane, but I haven't looked into how flexible the default Go date library is (I think it's not as flexible). It looks like there is a Go library for "natural" dates here: https://github.com/tj/go-naturaldate

File: internal/search/query/predicate.go (hunk 2 above)
Comment (rvantonder) [reply internal/search/query/predicate.go:163]: Gotcha, yeah, thanks for reminding me. I suppose I'm coming back to this because there might be a nicer midway than always running `99999`,  like "choose some value that's 500 or 1000 at first and then prompt the user to increase count and inherit that count value and if that count value then is `99999` so be it". It's basically a one-shot retry attempt. I think the retry loop would be a bit of effort to do now, and the one-shot approach would need investigating, it's a bit unpredictable how this would behave, so I'm OK with this.

File: internal/search/query/predicate.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (camdencheek) [reply internal/search/query/predicate.go:151]: I don't think this should be an immediate concern for this PR since we don't do any validation for `before`, `after`, and `repohascommitafter` currently, so there will be no regression. 

That said, I'd much prefer having a date format that we control here. I'd like to be able to validate and store it as a date rather than just an opaque string. I opened an issue [here](https://github.com/sourcegraph/sourcegraph/issues/19687) for tracking that.

File: internal/search/query/predicate.go (hunk 2 above)
Comment (camdencheek) [reply internal/search/query/predicate.go:163]: Created a followup issue here: https://github.com/sourcegraph/sourcegraph/issues/19688
